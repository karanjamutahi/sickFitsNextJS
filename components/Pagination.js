import React, { Component } from 'react';
import gql from 'graphql-tag';
import Link from 'next/link';
import Head from 'next/head';

import PaginationStyles from './styles/PaginationStyles';
import { Query } from 'react-apollo';
import Loading from './Loading';
import Error from './ErrorMessage';
import { perPage } from '../config';
import Title from './styles/Title';



const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        itemsConnection {
            aggregate {
                count
            }
        }
    }
`;


const Pagination = (props) => (
        <Query query={ PAGINATION_QUERY } >
            {({ data , loading, error }) => {
                if (loading) return <Loading loading={loading}></Loading>;
                if (error) return <Error error={error}></Error>; 
                const count = data.itemsConnection.aggregate.count;
                const pages = Math.ceil(count/perPage);
                let page = props.page;
                return (   
                <PaginationStyles>
                        <Head>
                            <title>Sick Fits | Page {page}</title>
                        </Head>
                        <Link 
                        prefetch
                        href={{
                            pathname:'items',
                            query: {
                                page: page>1 ? page -1 : pages
                            }
                        }}><a className='prev'>â—„ Prev</a></Link>
                        <p>{`Page ${page} of ${pages} ðŸ˜Ž`}</p>
                        <Link 
                        prefetch
                        href={{
                            pathname:'items',
                            query: {
                                page: page<pages ? page +1 : 1
                            }
                        }}><a>Next â–º</a></Link>
                </PaginationStyles>
                );
            }}
        </Query>
)

export default Pagination ;