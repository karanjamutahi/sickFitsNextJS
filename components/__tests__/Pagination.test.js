import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import Pagination, { PAGINATION_QUERY } from '../Pagination';
import { MockedProvider } from 'react-apollo/test-utils';

Router.router = {
  push() {},
  prefetch() {},
};

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

describe('Pagination Component', () => {
  it('displays a loading message', () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    //console.log(wrapper.debug());
    const pagination = wrapper.find('[data-test="pagination"]');
    expect(wrapper.find('Loading').exists()).toBeTruthy();
  });

  it('renders pagination for 18 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    //console.log(wrapper.debug());
    expect(wrapper.find('.totalPages').text()).toEqual('5');
    const pagination = wrapper.find('div[data-test="pagination"]');
    expect(toJSON(pagination)).toMatchSnapshot();
  });

});