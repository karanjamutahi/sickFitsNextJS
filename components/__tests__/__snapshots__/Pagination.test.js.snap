// Jest Snapshot v1, https://goo.gl/fbAQLP
exports[`Pagination Component renders pagination for 18 items 1`] = `
<div
  className="PaginationStyles-aduuar-0 ewJsCc"
  data-test="pagination"
>
  <Head>
    <Component
      handleStateChange={null}
      inAmpMode={false}
      reduceComponentsToState={[Function]}
    />
  </Head>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 5,
        },
      }
    }
    prefetch={true}
  >
    <a
      className="prev"
      href="items?page=5"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      â—„ Prev
    </a>
  </Link>
  <p>
    Page 
    1
     of 
    <span
      className="totalPages"
    >
      5
    </span>
     ðŸ˜Ž}
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      href="items?page=2"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Next â–º
    </a>
  </Link>
</div>
`;
